crates/safe/src/safe.nr

```diff
@@ -0,0 +1,700 @@
+            if self.absorb_pos == RATE {
+                // n-c = RATE.
+                self.state = self.permute();
+                self.absorb_pos = 0;
+            }
+
+            // Add X[i] to state at absorb_pos (spec 2.4).
+            // Note: absorb_pos is the rate position, not capacity position.
+            self.state[self.absorb_pos + CAPACITY] =
+                self.state[self.absorb_pos + CAPACITY] + input[i];
+            self.absorb_pos += 1;
+        }
+
+        // Verify that the encoded word matches the expected pattern.
+        let encoded_word = ABSORB_FLAG | length;
+        assert(encoded_word == expected_encoded_word);
+
+        self.io_count += 1;
+
+        // Force permute at start of next SQUEEZE (spec 2.4).
+        self.squeeze_pos = RATE;
+    }
+
+    /// Extracts field elements from the sponge state, interleaving permutation calls as needed (following spec 2.4).
+    ///
+    /// The number of elements to squeeze is automatically determined from the IO pattern.
+    ///
+    /// # Returns
+    /// Tuple of (output array, updated sponge instance).
+    /// - `output`: Vector of Field elements.
+    pub fn squeeze(&mut self) -> Vec<Field> {
+        // Parse expected operation from io_pattern (encoded word)
+        let expected_encoded_word = self.io_pattern[self.io_count];
+        let is_expected_squeeze = (expected_encoded_word & ABSORB_FLAG) == 0;
+        let length = expected_encoded_word & 0x7FFFFFFF;
+
+        // Validate operation type
+        assert(is_expected_squeeze, "Expected SQUEEZE operation");
+
+        let mut output = Vec::new();
+
+        // SQUEEZE implementation following spec 2.4.
+        // If length==0, loop won't execute (spec 2.4).
+        for _ in 0..length {
+            // If squeeze_pos==(n-c) then permute and reset (spec 2.4).
+            if self.squeeze_pos == RATE {
+                // n-c = RATE.
+                self.state = self.permute();
+                self.squeeze_pos = 0;
+                self.absorb_pos = 0;
+            }
+            // Set Y[i] to state element at squeeze_pos (spec 2.4).
+            output.push(self.state[self.squeeze_pos + CAPACITY]);
+            self.squeeze_pos += 1;
+        }
+
+        // Verify that the encoded word matches the expected pattern.
+        let encoded_word = SQUEEZE_FLAG | length;
+        assert(encoded_word == expected_encoded_word);
+
+        self.io_count += 1;
+        output
+    }
+
+    /// Finalizes the sponge instance, verifying that all expected operations have been performed and clearing the internal state for security (following spec 2.4).
+    ///
+    /// This function is used to ensure that the sponge instance has been used correctly and to prevent information leakage.
+    ///
+    /// # Returns
+    /// `true` if all operations in the IO pattern were completed successfully, `false` otherwise.
+    pub fn finish(&mut self) -> bool {
```

Comment:
I don't think you have to return a bool here; you can just have an assertion in the finish function and return nothing.

```suggestion
    pub fn finish(&mut self) {
```

Reply:
agree!